File: backend/app/services/nutrition.py
"""Расчёт нормы калорий и макросов (Миффлин–Сан Жеор)."""
from datetime import date

from app.models import ActivityLevel, Gender, Goal

ACTIVITY_FACTOR = {
    ActivityLevel.sedentary: 1.2,
    ActivityLevel.light: 1.375,
    ActivityLevel.moderate: 1.55,
    ActivityLevel.active: 1.725,
    ActivityLevel.very_active: 1.9,
}

GOAL_FACTOR = {
    Goal.lose: 0.82,      # −18%
    Goal.maintain: 1.0,
    Goal.gain: 1.12,      # +12%
}

# Распределение макросов (доля от калорий): белок / жир / углеводы
MACRO_SPLIT = {"protein": 0.30, "fat": 0.30, "carb": 0.40}
KCAL_PER_G = {"protein": 4, "fat": 9, "carb": 4}


def calc_bmr(gender: Gender, weight_kg: float, height_cm: float, age: int) -> float:
    base = 10 * weight_kg + 6.25 * height_cm - 5 * age
    return base + 5 if gender == Gender.male else base - 161


def calc_targets(
    gender: Gender,
    weight_kg: float,
    height_cm: float,
    birth_year: int,
    activity: ActivityLevel,
    goal: Goal,
) -> dict[str, int]:
    """Возвращает целевые калории и макросы (граммы)."""
    age = date.today().year - birth_year
    bmr = calc_bmr(gender, weight_kg, height_cm, age)
    tdee = bmr * ACTIVITY_FACTOR[activity]
    calories = round(tdee * GOAL_FACTOR[goal])

    return {
        "calorie_target": calories,
        "protein_target": round(calories * MACRO_SPLIT["protein"] / KCAL_PER_G["protein"]),
        "fat_target": round(calories * MACRO_SPLIT["fat"] / KCAL_PER_G["fat"]),
        "carb_target": round(calories * MACRO_SPLIT["carb"] / KCAL_PER_G["carb"]),
    }

File: miniapp/src/lib/nutrition.ts
/** Клиентский расчёт нормы (зеркало backend/app/services/nutrition.py). */

export type Gender = "male" | "female";
export type GoalId = "lose" | "maintain" | "gain";
export type ActivityId = "sedentary" | "light" | "moderate" | "active" | "very_active";

export const ACTIVITIES: { id: ActivityId; factor: number; label: string; desc: string }[] = [
  { id: "sedentary", factor: 1.2, label: "Минимальная", desc: "Сидячий образ жизни" },
  { id: "light", factor: 1.375, label: "Лёгкая", desc: "1–3 тренировки в неделю" },
  { id: "moderate", factor: 1.55, label: "Умеренная", desc: "3–5 тренировок в неделю" },
  { id: "active", factor: 1.725, label: "Высокая", desc: "6–7 тренировок в неделю" },
  { id: "very_active", factor: 1.9, label: "Очень высокая", desc: "Спорт + физическая работа" },
];

export const GOALS: { id: GoalId; factor: number; label: string; desc: string }[] = [
  { id: "lose", factor: 0.82, label: "Похудеть", desc: "Дефицит калорий" },
  { id: "maintain", factor: 1.0, label: "Поддерживать", desc: "Держать текущий вес" },
  { id: "gain", factor: 1.12, label: "Набрать", desc: "Профицит калорий" },
];

const MACRO_SPLIT = { protein: 0.3, fat: 0.3, carb: 0.4 };
const KCAL_PER_G = { protein: 4, fat: 9, carb: 4 };

export interface Anketa {
  gender: Gender;
  age: number;
  height_cm: number;
  weight_kg: number;
  activity: ActivityId;
  goal: GoalId;
}

export function calcTargets(a: Anketa) {
  const bmr =
    10 * a.weight_kg + 6.25 * a.height_cm - 5 * a.age + (a.gender === "male" ? 5 : -161);
  const factor = ACTIVITIES.find((x) => x.id === a.activity)!.factor;
  const goalFactor = GOALS.find((x) => x.id === a.goal)!.factor;
  const kcal = Math.round(bmr * factor * goalFactor);

  return {
    kcal,
    protein: Math.round((kcal * MACRO_SPLIT.protein) / KCAL_PER_G.protein),
    fat: Math.round((kcal * MACRO_SPLIT.fat) / KCAL_PER_G.fat),
    carb: Math.round((kcal * MACRO_SPLIT.carb) / KCAL_PER_G.carb),
  };
}
